notebooks/05.log_and_register_fe_model.py

```diff
@@ -0,0 +1,239 @@
+train_set = train_set.drop("housing", "loan", "default")
+
+# CAST primary key to "string"
+train_set = train_set.withColumn(
+    "Id", train_set["Id"].cast("string")
+)  # <- make sure "Id" it's string type!  # <- make sure "Id" it's string type!
+
+###########################
+# Feature engineering setup
+###########################
+
+# Create training_set with feature_engineering client fe
+training_set = fe.create_training_set(
+    df=train_set,
+    label=target,
+    feature_lookups=[
+        FeatureLookup(
+            table_name=feature_table_name,
+            feature_names=["default_", "housing", "loan"],
+            lookup_key="Id",
+        ),
+        FeatureFunction(
+            udf_name=function_name,
+            output_name="risk",
+            input_bindings={"default_": "default_", "housing": "housing", "loan": "loan"},
+        ),
+    ],
+    exclude_columns=["update_timestamp_utc"],
+)
+
+# Load feature-engineered DataFrame
+training_df = training_set.load_df().toPandas()
+training_df.rename(columns={"default_": "default"}, inplace=True)
+
+# Calculate house_age for training and test set
+# test_set = spark.createDataFrame(test_set)
+test_set = test_set.selectExpr("*", f"{function_name}(default, housing, loan) as risk")
+test_set = test_set.toPandas()
+
+# Split features and target
+num_features = list(training_df.select_dtypes("number").columns)
+cat_features = list(training_df.select_dtypes("category").columns) + ["risk"]
+X_train = training_df[num_features + cat_features]
+y_train = training_df[target]
+X_test = test_set[num_features + cat_features]
+y_test = test_set[target]
+
+########################################
+# Setup preprocessing and model pipeline
+########################################
+# Numeric Transformer
+numeric_transformer = Pipeline(steps=[("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())])
+
+# Categorical Transformer
+categorical_transformer = Pipeline(
+    steps=[
+        ("imputer", SimpleImputer(strategy="constant", fill_value="missing")),
+        ("onehot", OneHotEncoder(max_categories=5, handle_unknown="ignore")),
+    ]
+)
+
+# Label Encoder for target variable
+le = LabelEncoder()
+
+# Define the preprocessor
+preprocessor = ColumnTransformer(
+    transformers=[("num", numeric_transformer, num_features), ("cat", categorical_transformer, cat_features)],
+    remainder="passthrough",
+)
+
+# Create the pipeline with preprocessing and the RandomForestClassifier
+pipeline = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", RandomForestClassifier(**parameters))])
+
+##########
+# MLFlow #
+##########
+
+# Set and start MLflow experiment
+mlflow.set_experiment(experiment_name="/Shared/bank-marketing-fe")
+git_sha = "ffa63b430205ff7"
+
+with mlflow.start_run(tags={"branch": "week2", "git_sha": f"{git_sha}"}) as run:
+    run_id = run.info.run_id
+
+    # Pipeline fit/predict
+    pipeline.fit(X_train, le.fit_transform(y_train))
```

Comment:
you create the label encoder, but do not save the mapping anywhere. So you run into problems in the future, should not be that part of preprocessing?